engine/clientdb/dbs.go

```diff
@@ -17,61 +19,130 @@ import (
 
 type DBs struct {
 	Root string
+
+	open map[string]*DB
+	mu   sync.RWMutex
 }
 
 // CollectGarbageAfter is the time after which a database is considered
 // garbage and can be deleted.
 const CollectGarbageAfter = time.Hour
 
 func NewDBs(root string) *DBs {
-	return &DBs{Root: root}
+	return &DBs{
+		Root: root,
+		open: make(map[string]*DB),
+	}
 }
 
 //go:embed schema.sql
 var Schema string
 
-// Create creates a new database for the given clientID and runs the schema
-// migration. This operation must be idempotent.
-func (dbs *DBs) Create(clientID string) (*sql.DB, error) {
-	db, err := dbs.Open(clientID)
-	if err != nil {
-		return nil, err
+// Open open a database for the given clientID if not open already and
+// runs the schema migration if needed.
+func (dbs *DBs) Open(ctx context.Context, clientID string) (_ *DB, rerr error) {
+	lg := slog.Default().With("clientID", clientID)
+
+	dbs.mu.Lock()
+	db, ok := dbs.open[clientID]
+	if !ok {
+		db = &DB{
+			dbs:      dbs,
+			clientID: clientID,
+		}
+		dbs.open[clientID] = db
 	}
-	if _, err := db.Exec(Schema); err != nil {
-		return nil, fmt.Errorf("migrate: %w", err)
+	dbs.mu.Unlock()
+
+	db.mu.Lock()
+	defer db.mu.Unlock()
+	defer func() {
+		if rerr != nil {
+			rerr = errors.Join(rerr, dbs.close(db, lg))
+		}
+	}()
+	db.refCount++
+
+	lg = lg.With("aquiredRefCount", db.refCount)
+
+	if db.inner == nil {
+		lg.ExtraDebug("opening client DB", "clientID", clientID)
+
+		dbPath := db.dbs.path(db.clientID)
+		if err := os.MkdirAll(filepath.Dir(dbPath), 0700); err != nil {
+			return nil, fmt.Errorf("mkdir %s: %w", filepath.Dir(dbPath), err)
+		}
+		connURL := &url.URL{
+			Scheme: "file",
+			Host:   "",
+			Path:   dbPath,
+			RawQuery: url.Values{
+				"_pragma": []string{
+					"foreign_keys=ON",    // we don't use em yet, but makes sense anyway
+					"journal_mode=WAL",   // readers don't block writers and vice versa
+					"synchronous=OFF",    // we don't care about durability and don't want to be surprised by syncs
+					"busy_timeout=10000", // wait up to 10s when there are concurrent writers
+				},
+				"_txlock": []string{"immediate"}, // use BEGIN IMMEDIATE for transactions
+			}.Encode(),
+		}
+		sqlDB, err := sql.Open("sqlite", connURL.String())
+		if err != nil {
+			return nil, fmt.Errorf("open %s: %w", connURL, err)
+		}
+		if err := sqlDB.Ping(); err != nil {
+			return nil, fmt.Errorf("ping %s: %w", connURL, err)
+		}
+
+		db.inner = sqlDB
+		if _, err := db.inner.Exec(Schema); err != nil {
```

Comment:
nit: Seems like this will re-run the migrations when reopening an existing DB too, which is fine at the moment because they're idempotent, but may be a risk in the future. Could be worth a quick `os.Stat` just to be extra clear about the intention, but not blocking